### PR TITLE
Build: migrate from pipenv to poetry

### PR DESCRIPTION
**Pull request by roberta-dt**
# Description

Moved from `pipenv` to `poetry` for Python dependency management, and upgraded the Python version to 3.9 ahead of Python 3.7 EOL

# How has this been tested?

`make` commands tested manually
Automated tests all passing

# Checklist

~~- [ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have successfully run the E2E tests
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] New and existing unit tests pass locally with my changes
- [ ] I have updated any relevant documentation to reflect my changes
- [x] I have assigned a reviewer and messaged them

### CODE REVIEW

File: components/bigquery-components/pyproject.toml [hunk 1]
```diff
@@ -1,30 +1,30 @@
-[project]
+[tool.poetry]
 name = "bigquery-components"
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
-description = "BigQuery components"
+version = "0.1.0"
+authors = ["roberta nwokonko <roberta.nwokonko@datatonic.com>"]
```
Comment (browningjp-datatonic) [inline components/bigquery-components/pyproject.toml:4]: change author to an example name and email address

File: components/vertex-components/pyproject.toml [hunk 2]
```diff
@@ -1,32 +1,28 @@
-[project]
+[tool.poetry]
 name = "vertex-components"
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
-description = "Vertex AI components"
+version = "0.1.0"
+authors = ["roberta nwokonko <roberta.nwokonko@datatonic.com>"]
```
Comment (browningjp-datatonic) [inline components/vertex-components/pyproject.toml:4]: change author to example

File: components/bigquery-components/pyproject.toml [hunk 3]
```diff
@@ -1,30 +1,30 @@
...
+description = "BigQuery components created by Datatonic"
```
Comment (browningjp-datatonic) [inline components/bigquery-components/pyproject.toml:5]: remove references to Datatonic - instead something like "BigQuery components for Vertex Pipelines"

File: pipelines/pyproject.toml [hunk 4]
```diff
@@ -1,30 +1,28 @@
-[project]
+[tool.poetry]
 name = "turbo-template"
```
Comment (browningjp-datatonic) [inline pipelines/pyproject.toml:2]: rename to `pipelines`

File: pipelines/pyproject.toml [hunk 5]
```diff
@@ -1,30 +1,28 @@
...
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
-description = "Turbo Template"
+version = "0.1.0"
+authors = ["roberta nwokonko <roberta.nwokonko@datatonic.com>"]
```
Comment (browningjp-datatonic) [inline pipelines/pyproject.toml:4]: use example user

File: poetry.lock
Comment (browningjp-datatonic) [inline poetry.lock:1]: I don't think we need `poetry.lock` and `pyproject.toml` in the root of the repo

**Review by ghost — CHANGES_REQUESTED** [review]
Remember also to update the CI/CD pipelines in the `cloudbuild` directory to use a `python:3.9` image

File: poetry.lock
Comment (roberta-dt) [reply poetry.lock:1]: 7813fc7

File: pipelines/pyproject.toml (hunk 4 above)
Comment (roberta-dt) [reply pipelines/pyproject.toml]: 66c0da1

File: pipelines/pyproject.toml [hunk 6]
```diff
@@ -1,30 +1,28 @@
...
+description = "Turbo Template developed at Datatonic"
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply pipelines/pyproject.toml:7]: 265777f

File: pipelines/pyproject.toml (hunk 5 above)
Comment (roberta-dt) [reply pipelines/pyproject.toml]: 265777f

File: components/vertex-components/pyproject.toml [hunk 7]
```diff
@@ -1,32 +1,28 @@
...
+description = "Vertex AI components made by Datatonic"
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml]: 265777f

File: components/bigquery-components/pyproject.toml (hunk 3 above)
Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: 265777f

File: components/vertex-components/pyproject.toml (hunk 2 above)
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml]: 265777f

File: components/bigquery-components/pyproject.toml (hunk 1 above)
Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: 265777f

File: components/vertex-components/pyproject.toml [hunk 8]
```diff
@@ -1,32 +1,30 @@
-[project]
+[tool.poetry]
 name = "vertex-components"
+version = "0.1.0"
 authors = [
     {name = "Example User", email = "user@example.com"},
 ]
-description = "Vertex AI components"
+description = "Vertex components for training and prediction using the Chicago Taxi Dataset"
```
Comment (browningjp-datatonic) [inline components/vertex-components/pyproject.toml:7]: these components aren't specific to the usecase - maybe description should be something like "KFP components for interacting with Vertex AI"

File: pipelines/pyproject.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (browningjp-datatonic) [reply pipelines/pyproject.toml:7]: not sure what happened here - the commit you have linked to is not on this branch

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by browningjp-datatonic** [comment]
@roberta-dt you will also need to update the Makefile to change all the `pipenv` commands to `poetry` commands

**Comment by roberta-dt** [comment]
\gcbrun

File: components/bigquery-components/pyproject.toml [hunk 9]
```diff
@@ -1,30 +1,30 @@
...
+python = ">=3.9,<3.11"
+Jinja2 = ">=3.0.1,<4.0.0"
+google-cloud-aiplatform = "1.24.1"
+google-cloud-pipeline-components = "1.0.42"
+google-cloud-bigquery = "2.30.0"
+kfp = "==1.8.21"
+
+[tool.poetry.group.dev.dependencies]
+pytest = ">=7.3.1,<8.0.0"
+pre-commit = ">=2.14.1,<3.0.0"
```
Comment (browningjp-datatonic) [inline components/bigquery-components/pyproject.toml:23]: the dependencies should match the previous dependencies from `pyproject.toml`:

```
[tool.poetry.dependencies]
python = ">=3.9,<3.11"
kfp = "==1.8.21"

[tool.poetry.group.dev.dependencies]
google-cloud-bigquery = "2.30.0"
pytest = ">=7.3.1,<8.0.0"
```

File: components/vertex-components/pyproject.toml [hunk 10]
```diff
@@ -1,32 +1,28 @@
...
-]
+[tool.poetry.dependencies]
+python = ">=3.9,<3.11"
+google-cloud-aiplatform = "1.24.1"
+google-cloud-pipeline-components = "1.0.42"
+kfp = "==1.8.21"
+
+[tool.poetry.group.dev.dependencies]
+pytest = ">=7.3.1,<8.0.0"
+pre-commit = ">=2.14.1,<3.0.0"
```
Comment (browningjp-datatonic) [inline components/vertex-components/pyproject.toml:21]:
```
[tool.poetry.dependencies]
python = ">=3.9,<3.11"
kfp = "==1.8.21"

[tool.poetry.group.dev.dependencies]
google-cloud-aiplatform = "1.24.1"
google-cloud-pipeline-components = "1.0.42"
pytest = ">=7.3.1,<8.0.0"
```

File: pipelines/pyproject.toml [hunk 11]
```diff
@@ -1,30 +1,37 @@
...
-    "google-cloud-bigquery == 2.30.0",
-    "pytest >= 7.3.1,<8.0.0",
-]
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.11"
+Jinja2 = ">=3.0.1,<4.0.0"
+google-cloud-aiplatform = "1.24.1"
+google-cloud-pipeline-components = "1.0.42"
+pytest = "^7.3.1"
```
Comment (browningjp-datatonic) [inline pipelines/pyproject.toml:22]: remove `pytest` from dependencies (as it is already in the dev dependencies below)

**Review by ghost — CHANGES_REQUESTED** [review]

File: components/vertex-components/pyproject.toml (hunk 10 above)
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml]: cec9f3e

File: components/bigquery-components/pyproject.toml (hunk 9 above)
Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: cec9f3e

Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: cec9f3e

File: pipelines/pyproject.toml (hunk 11 above)
Comment (roberta-dt) [reply pipelines/pyproject.toml]: 149ab44

File: components/bigquery-components/pyproject.toml (hunk 9 above)
Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: Both bigquery and vertex components were squashed together in this commit

**Comment by roberta-dt** [comment]
\gcbrun

File: components/vertex-components/pyproject.toml (hunk 10 above)
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml]: Both bigquery and vertex components were squashed together in this commit

**Review by ghost — CHANGES_REQUESTED** [review]
There are a few places where `pipenv` is still mentioned in the docs or in comments - see the search results [here](https://github.com/search?q=repo%3Ateamdatatonic%2Fvertex-pipelines-end-to-end-samples%20pipenv&type=code). Please could you update these?

**Comment by browningjp-datatonic** [comment]
> There are a few places where `pipenv` is still mentioned in the docs or in comments - see the search results [here](https://github.com/search?q=repo%3Ateamdatatonic%2Fvertex-pipelines-end-to-end-samples%20pipenv&type=code). Please could you update these?

Likewise for any mentions of `Pipfile` in the docs e.g. in [components/README.md](components/README.md)

File: README.md [hunk 12]
```diff
@@ -55,8 +55,8 @@ In a production MLOps solution, your ML pipelines need to be repeatable. So, we
 
 1. Clone the repository locally
 1. Install Python: `pyenv install`
-1. Install pipenv and pipenv dependencies: `make setup`
-1. Install pre-commit hooks: `cd pipelines && pipenv run pre-commit install`
+1. Install poetry and  dependencies: `make setup`
+1. Install pre-commit hooks: `cd pipelines` && run `pre-commit install`
```
Comment (browningjp-datatonic) [inline README.md:59]: This line should read

> 1. Install pre-commit hooks: `cd pipelines && poetry run pre-commit install`

**Review by ghost — DISMISSED** [review]

File: README.md (hunk 12 above)
Comment (roberta-dt) [reply README.md]: bcea327

**Comment by browningjp-datatonic** [comment]
I think we need to improve some of the setup instructions relating to pyenv and poetry - and possibly the `make setup` command:

- Add instruction to install `pyenv`
- Add instruction to install `poetry` (minimum version tbd for next step) - [Poetry docs](https://python-poetry.org/docs/#installation) give a few options for installing `poetry` and warn against installing directly with `pip`
- Add instruction for configuring Poetry to use the active Python version from `pyenv` - `poetry config virtualenvs.prefer-active-python true`
- `poetry install` etc...

We also need to rebase on top of `develop` before this can be merged

@felix-datatonic please can you support with this in my absence?

**Comment by roberta-dt** [comment]
> I think we need to improve some of the setup instructions relating to pyenv and poetry - and possibly the `make setup` command:
> 
> * Add instruction to install `pyenv`
> * Add instruction to install `poetry` (minimum version tbd for next step) - [Poetry docs](https://python-poetry.org/docs/#installation) give a few options for installing `poetry` and warn against installing directly with `pip`
> * Add instruction for configuring Poetry to use the active Python version from `pyenv` - `poetry config virtualenvs.prefer-active-python true`
> * `poetry install` etc...
> 
> We also need to rebase on top of `develop` before this can be merged
> 
> @felix-datatonic please can you support with this in my absence?

Okay. I made some setup changes in the README.MD files yesterday on how to install poetry and install components as packages, I am not sure if you can see those changes yet, it may be because of the conflicts on the branch. I'll rebase and add the rest of changes.

**Comment by ariadnafer** [comment]
I think we need to change also the python version the base images of all the components, no? @roberta-dt

**Comment by roberta-dt** [comment]
> I think we need to change also the python version the base images of all the components, no? However, checking here: https://cloud.google.com/vertex-ai/docs/training/pre-built-containers. it seems we only have available python 3.7 and 3.10 for the training & predictions images, thoughts?
> 
Yes, I think so

File: .python-version [hunk 13]
```diff
@@ -1 +1 @@
-3.7.12
+3.7.12
```
Comment (felix-datatonic) [inline .python-version:1]:
```suggestion
3.7.12
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -211,12 +211,12 @@ def test_vertex_endpoint_uri(output_uri: str):
 ```
 
 ## Adding or changing python dependencies
-We use [pipenv](https://pipenv-fork.readthedocs.io/en/latest/index.html) to handle our packages and their dependencies. Each group of pipeline components (e.g. [aiplatform](./pipeline_components/aiplatform/)) containers its own pipenv environment, and there is a [separate pipenv environment](./pipelines/) for the ML pipelines themselves and the pipeline trigger code.
+We use [poetry](https://python-poetry.org/docs/#installation) to handle our packages and their dependencies. Each group of pipeline components (e.g. [aiplatform](./pipeline_components/aiplatform/)) containers its own poetry environment, and there is a [separate poetry environment](./pipelines/) for the ML pipelines themselves and the pipeline trigger code.
```
Comment (felix-datatonic) [inline CONTRIBUTING.md:214]:
```suggestion
We use [poetry](https://python-poetry.org/docs/#installation) to handle our packages and their dependencies. Each group of pipeline components (e.g. [vertex](./components/vertex-components/)) includes its own poetry environment, and there is a [separate poetry environment](./pipelines/) for the ML pipelines themselves and the pipeline trigger code.
```

File: Makefile [hunk 15]
```diff
@@ -19,25 +19,25 @@ help: ## Display this help screen
...
-	pipenv run pre-commit run --all-files
+	cd pipelines && \
+	poetry run pre-commit run --all-files
 
 setup: ## Set up local environment for Python development on pipelines
-	@pip install pipenv && \
+	@pip install poetry && \
 	cd pipelines && \
-	pipenv install --dev
+	poetry install --with dev
```
Comment (felix-datatonic) [inline Makefile:28]:
```suggestion
	poetry install --with dev
```

File: pipelines/pyproject.toml [hunk 16]
```diff
@@ -1,31 +1,33 @@
-[project]
-name = "turbo-template"
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
+[tool.poetry]
+name = "pipelines"
+version = "0.1.0"
+authors = ["Example User <user@example.com>"]
 description = "Turbo Template"
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply pipelines/pyproject.toml]: ee5bb52

File: components/bigquery-components/pyproject.toml [hunk 17]
```diff
@@ -1,30 +1,26 @@
-[project]
+[tool.poetry]
 name = "bigquery-components"
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
-description = "BigQuery components"
+version = "0.1.0"
+authors = ["Example User <user@example.com>"]
+description = "Bigquery components for vertex pipelines"
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/bigquery-components/pyproject.toml]: 43463f8

File: README.md [hunk 18]
```diff
@@ -55,11 +55,18 @@ In a production MLOps solution, your ML pipelines need to be repeatable. So, we
...
+1. Install pre-commit hooks: `cd pipelines && poetry run pre-commit install`
 1. Copy `env.sh.example` to `env.sh`, and update the environment variables in `env.sh`
 1. Load the environment variables in `env.sh` by running `source env.sh`
 
+please note: `poetry install` or `poetry add`, installs packages within the project's virtual environment. If you use pip directly, you might accidentally install packages globally or in the wrong environment, leading to conflicts or difficulties in managing dependencies.
+
+### Configuring poetry to detect python version using pyenv
+
+1. run in terminal `poetry config virtualenvs.prefer-active-python true`
+1. Manage and install project's dependencies using `poetry install`
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply README.md]: bf2fa78

File: CONTRIBUTING.md (hunk 14 above)
Comment (roberta-dt) [reply CONTRIBUTING.md]: e39dc49

File: components/vertex-components/pyproject.toml [hunk 19]
```diff
@@ -1,32 +1,27 @@
-[project]
+[tool.poetry]
 name = "vertex-components"
-authors = [
-    {name = "Example User", email = "user@example.com"},
-]
-description = "Vertex AI components"
+version = "0.1.0"
+authors = ["Example User <user@example.com>"]
+description = "KFP components for interacting with Vertex AI"
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml:5]: 4adc66e

[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml:5]: 4adc66e

[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml:5]: 4adc66e

File: .python-version (hunk 13 above)
Comment (roberta-dt) [reply .python-version]: ebcc591

File: components/vertex-components/.python-version [hunk 20]
```diff
@@ -1 +1 @@
-3.7.12
+3.7.12
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/.python-version]: ebcc591

File: pipelines/.python-version [hunk 21]
```diff
@@ -1 +1 @@
-3.7.12
+3.7.12
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply pipelines/.python-version]: ebcc591

File: Makefile (hunk 15 above)
Comment (roberta-dt) [reply Makefile]: 3a694cd

File: Makefile [hunk 22]
```diff
@@ -19,25 +19,25 @@ help: ## Display this help screen
...
 
 compile-pipeline: ## Compile the pipeline to training.json or prediction.json. Must specify pipeline=<training|prediction>
 	@cd pipelines/src && \
-	pipenv run python -m pipelines.${PIPELINE_TEMPLATE}.${pipeline}.pipeline
+	poetry run python -m pipelines.${PIPELINE_TEMPLATE}.${pipeline}.pipeline
 
 setup-components: ## Run unit tests for a component group
 	@cd "components/${GROUP}" && \
-	pipenv install --dev
+	poetry install --with dev
```
[the remark this answers is not in the document]
Comment (roberta-dt) [reply Makefile]: 3a694cd

File: components/vertex-components/pyproject.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (roberta-dt) [reply components/vertex-components/pyproject.toml]: 4adc66e

**Review by felix-datatonic — DISMISSED** [review]
Thanks a lot @roberta-dt for your contribution! Approved.

File: README.md [hunk 23]
```diff
@@ -55,11 +55,19 @@ In a production MLOps solution, your ML pipelines need to be repeatable. So, we
 
 1. Clone the repository locally
 1. Install Python: `pyenv install`
-1. Install pipenv and pipenv dependencies: `make setup`
-1. Install pre-commit hooks: `cd pipelines && pipenv run pre-commit install`
+1. Install poetry and poetry dependencies: `make setup`
```
Comment (ariadnafer) [inline README.md:58]: This is redundant, no?

File: pipelines/pyproject.toml [hunk 24]
```diff
@@ -1,31 +1,33 @@
...
+python = ">=3.7,<3.11"
+Jinja2 = ">=3.0.1,<4.0.0"
+google-cloud-aiplatform = "1.24.1"
+google-cloud-pipeline-components = "1.0.42"
+bigquery-components = { path = "../components/bigquery-components", develop = true }
+vertex-components = { path = "../components/vertex-components", develop = true }
+
+[tool.poetry.group.dev.dependencies]
+pytest = ">=7.3.1,<8.0.0"
+pre-commit = ">=2.14.1,<3.0.0"
```
Comment (ariadnafer) [inline pipelines/pyproject.toml:26]: Missing `"coverage = ==7.2.5"` package

**Review by ariadnafer — CHANGES_REQUESTED** [review]

File: README.md (hunk 23 above)
Comment (roberta-dt) [reply README.md:58]: No it's not, because in the Makefile we run pre-commit, but here we install it

File: pipelines/pyproject.toml (hunk 24 above)
Comment (roberta-dt) [reply pipelines/pyproject.toml:26]: I'll add using a hotfix, the PR has been merged.

File: README.md (hunk 23 above)
Comment (ariadnafer) [reply README.md:58]: You are right!

File: pipelines/pyproject.toml (hunk 24 above)
Comment (roberta-dt) [reply pipelines/pyproject.toml:26]: 874d282

Comment (roberta-dt) [reply pipelines/pyproject.toml:26]: Done

**Review by ariadnafer — APPROVED** [review]

**Review by felix-datatonic — APPROVED** [review]

**Pull request merged**